Add flatten_structure parameter to GCSToS3Operator (#56134)

HaJunYoo opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

closes: #56134

Add `flatten_structure` parameter to `GCSToS3Operator`

**Summary**

Adds `flatten_structure` parameter to `GCSToS3Operator` that removes directory structure from transferred files, uploading only the filename to  the S3 destination path.

**Use case/motivation**

Currently, GCSToS3Operator always preserves the full GCS object path when uploading to S3. 
This makes it impossible to reorganize file structure during transfer without creating intermediate buckets or complex workarounds.

**Example:**
  ```python
# Example: Multiple parquet files in nested directories
  # GCS structure:
  # - data/sales/2025/01/15/transactions.parquet
  # - data/sales/2025/01/15/customers.parquet
  # - data/sales/2025/01/16/transactions.parquet

  # Before (current behavior / as-is)
  GCSToS3Operator(
      gcs_bucket="my-bucket",
      prefix="data/sales/2025/01/15/",
      dest_s3_key="s3://target-bucket/processed/"
  )
  # Results in:
  # - s3://target-bucket/processed/data/sales/2025/01/15/transactions.parquet
  # - s3://target-bucket/processed/data/sales/2025/01/15/customers.parquet

  # After (with flatten_structure=True / to-be)
  GCSToS3Operator(
      gcs_bucket="my-bucket",
      prefix="data/sales/2025/01/15/",
      dest_s3_key="s3://target-bucket/processed/daily/",
      flatten_structure=True
  )
  # Results in:
  # - s3://target-bucket/processed/daily/transactions.parquet
  # - s3://target-bucket/processed/daily/customers.parquet
```

**Changes**

- Add `flatten_structure` parameter (default: False)
- Add path transformation with duplicate filename warnings
- Update documentation

**Test plan**

- Added comprehensive unit and integration tests
- Verified backward compatibility
- All existing tests pass

**Backward compatibility**

- Default `flatten_structure=False` preserves existing behavior
- All existing tests pass without modification


<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

HaJunYoo committed:
Add flatten_structure parameter to GCSToS3Operator

KoviAnusha commented on providers/amazon/src/airflow/providers/amazon/aws/transfers/gcs_to_s3.py:
```diff
@@ -79,6 +79,9 @@ class GCSToS3Operator(BaseOperator):
          on the bucket is recreated within path passed in dest_s3_key.
+    :param flatten_structure: (Optional) When set to True, places all files directly
+        in the dest_s3_key directory without preserving subdirectory structure.
+        Overrides keep_directory_structure when enabled.
```
Nice doc addition! Also, you can mention explicitly here that flatten_structure takes precedence over keep_directory_structure. Users often skim docstrings, so a one-liner note here helps avoid confusion later I believe.

HaJunYoo replied:
> Nice doc addition! Also, you can mention explicitly here that flatten_structure takes precedence over keep_directory_structure. Users often skim docstrings, so a one-liner note here helps avoid confusion later I believe.

Thanks for your thoughtful review.
Added explicit documentation about the `keep_directory_structure` parameter's precedence as suggested. [96a2a36](https://github.com/apache/airflow/commit/96a2a360906db8297eba377efec02ff36fecfac5) and [5fd0d92](https://github.com/apache/airflow/pull/57713/commits/5fd0d922773a312b954a8eb83d8ed6c2e821d3e9)

KoviAnusha replied:
Thanks for addressing the feedback. LGTM

KoviAnusha commented on providers/amazon/src/airflow/providers/amazon/aws/transfers/gcs_to_s3.py:
```diff
@@ -124,6 +128,10 @@ def __init__(
+        self.flatten_structure = flatten_structure
+
+        if self.flatten_structure and self.keep_directory_structure:
+            self.log.warning("flatten_structure=True overrides keep_directory_structure=True")
```
Good for safeguarding here. How about adding this precedence note to the top-level docstring so users immediately understand that flatten_structure=True overrides the directory flag?

HaJunYoo replied:
> Good for safeguarding here. How about adding this precedence note to the top-level docstring so users immediately understand that flatten_structure=True overrides the directory flag?

I've addressed your suggestion (adding note) in this commit
[96a2a36](https://github.com/apache/airflow/commit/96a2a360906db8297eba377efec02ff36fecfac5) and [5fd0d92](https://github.com/apache/airflow/pull/57713/commits/5fd0d922773a312b954a8eb83d8ed6c2e821d3e9)

KoviAnusha commented on providers/amazon/tests/unit/amazon/aws/transfers/test_gcs_to_s3.py:
```diff
@@ -330,6 +330,133 @@ def test_execute_without_keep_director_structure(self, mock_hook):
+            mock_log.warning.assert_called_once_with(expected_warning)
+
+    def test_transform_file_path_with_flatten_structure(self):
+        """Test _transform_file_path method with flatten_structure=True."""
```
I think parameterizing these file path checks with 'pytest.mark.parametrize' would make it even cleaner and easier to extend in the future.

HaJunYoo replied:
Thanks for the feedback.
Refactored tests with `pytest.mark.parametrize` in this commit [96a2a36](https://github.com/apache/airflow/commit/96a2a360906db8297eba377efec02ff36fecfac5)

HaJunYoo committed:
Enhance GCSToS3Operator docs clarity and refactor tests to use pytest.mark.parametrize for better maintainability

HaJunYoo committed:
Improve GCSToS3Operator documentation consistency and add concrete examples

  - Use consistent "takes precedence over" terminology in warning messages
  - Add specific file path transformation example in class docstring

HaJunYoo committed:
Fix GCSToS3Operator test log mocking for Airflow SDK compatibility

  - Replace mock.patch.object() with mock.patch() string path to handle
  - read-only log property in new Airflow SDK.
